psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineClient.java

```diff
@@ -0,0 +1,120 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.linqs.psl.application.inference.online;
+
+import org.linqs.psl.application.inference.online.messages.OnlineMessage;
+import org.linqs.psl.application.inference.online.messages.actions.controls.Exit;
+import org.linqs.psl.application.inference.online.messages.actions.controls.Stop;
+import org.linqs.psl.application.inference.online.messages.responses.OnlineResponse;
+import org.linqs.psl.config.Options;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.net.Socket;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * A client that communicates with an OnlineServer using OnlineMessages.
+ */
+public class OnlineClient implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(OnlineClient.class);
+
+    private List<OnlineResponse> serverResponses;
+    private BlockingQueue<OnlineMessage> actionQueue;
+    private String hostname;
+    private int port;
+
+    public OnlineClient(BlockingQueue<OnlineMessage> actionQueue, List<OnlineResponse> serverResponses) {
+        this.serverResponses = serverResponses;
+        this.actionQueue = actionQueue;
+        this.hostname = Options.ONLINE_HOST.getString();
+        this.port = Options.ONLINE_PORT_NUMBER.getInt();
+    }
+
+    public void run() {
+        OnlineMessage onlineAction = null;
+
+        try (
+                Socket server = new Socket(hostname, port);
+                ObjectOutputStream socketOutputStream = new ObjectOutputStream(server.getOutputStream());
+                ObjectInputStream socketInputStream = new ObjectInputStream(server.getInputStream())) {
+            // Startup serverConnectionThread for reading server responses.
+            ServerConnectionThread serverConnectionThread = new ServerConnectionThread(socketInputStream, serverResponses);
+            serverConnectionThread.start();
+
+            // Deque actions and send to server.
+            while (!(onlineAction instanceof Exit || onlineAction instanceof Stop)) {
+                // Dequeue next action.
+                try {
+                    onlineAction = actionQueue.take();
+                } catch (InterruptedException ex) {
+                    log.warn("Interrupted while taking an online action from the queue.", ex);
+                    continue;
```

Comment:
Unless there is a special case in the online system, you can just gracefully exit on an InterruptedException (treat it like an Exit) .
When shutting down, it's common to first interrupt all the other threads to give them a chance to shutdown before killing them.

(Same for the other interrupt.)